src/client/automation/playback/click/index.js

```diff
@@ -57,10 +59,17 @@ export default class ClickAutomation extends VisibleElementAutomation {
         eventUtils.bind(element, 'blur', onblur, true);
     }
 
+    // NOTE:
+    // If `touchstart`, `touchmove`, or `touchend` are canceled, we should not dispatch any mouse event
+    // that would be a consequential result of the prevented touch event
+    _isTouchEventWasCancelled () {
+        return this.eventState.touchStartCancelled || this.eventState.touchEndCancelled;
+    }
+
     _raiseTouchEvents (eventArgs) {
         if (featureDetection.isTouchDevice) {
-            eventSimulator.touchstart(eventArgs.element, eventArgs.options);
-            eventSimulator.touchend(eventArgs.element, eventArgs.options);
+            this.eventState.touchStartCancelled = !eventSimulator.touchstart(eventArgs.element, eventArgs.options);
+            this.eventState.touchEndCancelled   = !eventSimulator.touchend(eventArgs.element, eventArgs.options);
```

Comment:
The `touchend` action will happen even if we add `preventDefault` on `touchstart`?

Reply:
Yes, it should. As stated in the W3C specification:

> If, however, either the touchstart, touchmove or touchend event has been canceled during this interaction, no mouse or click events will be fired, and the resulting sequence of events would simply be:

> 1. touchstart
> 2. Zero or more touchmove events, depending on movement of the finger
> 3. touchend